### ISSUE TITLE
Vectorized explain_prediction (?)

### ISSUE THREAD

**Issue by lopuhin** [TeamHG-Memex/eli5#29]
It might be possible to improve explain_prediction performance for large number of items by vectorizing predictions and argmax calculations (see for example http://vmprof.com/#/d7e9c13357f50a5ee4d00a69200693f5?id=0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0&view=flames). But it's not clear if this is worth additional complications.

**Comment by kmike** [comment]
how bad is the current speed?

**Comment by lopuhin** [comment]
For explaining deep-deep link predictions with 2M features and top=1000 or 100 (almost the same speed) it's about 15 predictions/second, so it's not too bad.

**Comment by kmike** [comment]
Based on profile results vectorizing can give us up to 30 predictions/sec; it might be worth the hassle, but it depends on how complex are the code changes.

API also worths considering: eli5.lime explains a single prediction; with multiple predictions it could make sense to both explain them individually or use them all at once to explain a classifier in some way. But that's a moot point, and it can be solved with a flag later. So it really boils down to how hard is it to implement this feature and how more complex the code becomes.

**Comment by kmike** [comment]
Ah, it is not 30 predictions/sec, but likely a much smaller benefit, I misread the profiler results

**Comment by lopuhin** [comment]
I was thinking of something like 3x-5x improvement, because it seems to me that all operations in the profile have the potential to be sped up with vectorisation. But maybe the speedup is much less.

**Comment by kmike** [comment]
3x-5x improvement would be nice, I thought only about vectorizing argsort functions.

**Comment by lopuhin** [comment]
Preliminary tests (run at https://github.com/TeamHG-Memex/eli5/commit/09f4fb6c6f15191e2c19c5696aee0303476fbd5e#diff-a672d624d119711c1683ddb550ffc381R291) show that the vectorized variant is 2x slower, not faster! The profile for vectorized variant is here http://vmprof.com/#/44b211afff718a871b2cec5d5552feb9?id=0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0&view=flames and the loop before it (just running the same code but on individual documents) is here http://vmprof.com/#/a86d5a0821f3535984e27e89dd30361a?id=0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0&view=flames).
Hopefully I did nothing stupid with numpy here, I'll later checked the argpartition function (should have done it before, not after):
```
In [10]: %time for row in x: np.argpartition(row, -k)[-k:]
CPU times: user 532 ms, sys: 214 ms, total: 745 ms
Wall time: 757 ms

In [11]: %time indices = np.argpartition(x, -k, axis=1)[:, -k:]
CPU times: user 724 ms, sys: 584 ms, total: 1.31 s
Wall time: 1.41 s

In [12]: x.shape
Out[12]: (100, 1048577)
```

So indeed doing vectorized argpartition is 2x slower.

**Comment by lopuhin** [comment]
Initial experiments turned out to be not very promising, will close this and the branch (	b0ebc3, 09f4fb6)
